### DISCUSSION TITLE
FMU Export and Implementation in Simulink

### DISCUSSION THREAD

**Discussion by a-reiter** [OpenModelica/OpenModelica#11705] [Q&A] [answered]
Hi all,
I've recently generated a Modelica model (see attachment BatterySystem.mo), which I've exported to FMU (see attachment BatterySystem.fmu) using OpenModelica (v 1.22 64bit). I used the default FMI settings. When implementing this fmu into my Matlab Simulink (2022a) model, the following error occurred:

"Error in supplied FMU:  Unable to load dynamic library [...]\win64.BatterySystem.dll"

Also, the FMU checker gave "Could not load dll".

Unfortunately, I have no clue how to proceed and could not find anything of the kind online. I usually only work with FMpy in Python, which to this day always worked flawlessly.

I would be thankful for every helpful comment.

Best

Alex
[Attachments.zip](https://github.com/OpenModelica/OpenModelica/files/13621255/Attachments.zip)

**Comment by AnHeuermann** [comment, marked as answer]
There are three scenarios what could be the issue here:

  1. The compilation of the FMU failed, so there is no BatterySystem.ddl inside the FMU binary directory.
  2. The dynamic library is missing some dependencies to other dynamic libraries.
  3. You are using the FMU on an target it wasn't compiled for.

Check if you got any error messages from OpenModelica telling you the compilation failed.

Then to find your issue you can start by unzipping the FMU (e.g. using 7-Zip) and check if there is a file `FMU/binaries/win64/BatterySystem.dll` to see if you did run into case 1.
You can use tools like [Dependency](https://github.com/lucasg/Dependencies) to check why the dll can't be loaded.

**Reply by a-reiter** [reply]
Thank you very much for your input!
I proceeded as proposed and got the following results:

1. The FMU compilation in OpenModelica did not give any error message.
2. The BatterySystem.dll was existent inside the binaries/win64/ folder.

I further used Dependencies to load the dll. I've attached a screenshot of the result to this response. The two "sub"-dlls "libsundials_cvode.dll" and "libwinpthread-1.dll" were marked with red question marks and gave the message "We could not find [...] on the disk anymore [...]."
To me (who has absolutely no clue about dlls), this seems like a potential source for the fault. Changing the output solver in OMedit to Explicit Euler resolved the "libsundials_cvode.dll"-problem. However, the implementation in Simulink was still not successful. 

Do you have any suggestions, how I could resolve the error?
![Dependencies_cvode](https://github.com/OpenModelica/OpenModelica/assets/76156948/6b82d8fd-d67f-43ac-aa20-452945edc109)

**Reply by AnHeuermann** [reply]
Ah yes, that's a known issue: https://github.com/OpenModelica/OpenModelica/issues/11604.
There are some workarounds, but they are limited or not very user friendly, see https://github.com/OpenModelica/OpenModelica/issues/11604#issuecomment-1818514869

You can disable CVODE as CS solver and use explicit Euler instead.

![grafik](https://github.com/OpenModelica/OpenModelica/assets/38031952/22f48bba-34a1-492f-938e-2e1d20c45905)

If you are familiar with DLL dependencies you could also try to copy the missing dependencies manually and re-zip the FMU. They are stored in `C:\Program Files\OpenModelica1.22.0-64bit\bin` and `C:\Program Files\OpenModelica1.22.0-64bit\tools\msys\mingw64\bin`.

**Reply by a-reiter** [reply]
Love it! 
I use Explicit Euler for the export and copy the "libwinpthread-1.dll" into the same folder as the BatterySystem.dll. After rezipping and renaming in .fmu, it works flawlessly in Simulink.

Thank you very much!
